### PR TITLE
[JENKINS-47827] Pod and container runAsUser and runAsGroup

### PR DESCRIPTION
**Pull request by georgelee1**
This is to finish off the work started in https://github.com/jenkinsci/kubernetes-plugin/pull/583, this PR when merged can close 583.

Have added `runAsUser` and `runAsGroup` to both pod and containers so can set either at the pod level or more fine grain container level. Have also adjusted/added tests to ensure the yaml merge and template setting works correctly.

If the value is not set then the value will default to `null` which means that it is not set in the K8s deployment, which then means the user and group default to what ever the container image defined.

### CODE REVIEW

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 1]
```diff
@@ -506,6 +519,60 @@ public void yamlOverrideVolume() throws Exception {
...
+        );
+        child.setInheritFrom("parent");
+        child.setYamlMergeStrategy(merge());
+        setupStubs();
+        PodTemplate result = combine(parent, child);
+        Pod pod = new PodTemplateBuilder(result).withSlave(slave).build();
+        assertThat(pod.getSpec().getContainers(), hasSize(2));
+        Optional<Container> container = pod.getSpec().getContainers().stream().filter(c -> "container".equals(c.getName())).findFirst();
+        assertTrue(container.isPresent());
+        assertEquals(3000L, (Object) pod.getSpec().getSecurityContext().getRunAsUser());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(3000L, pod.getSpec().getSecurityContext().getRunAsUser());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 2]
```diff
@@ -506,6 +519,60 @@ public void yamlOverrideVolume() throws Exception {
...
+        assertEquals(3000L, (Object) pod.getSpec().getSecurityContext().getRunAsGroup());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(3000L, pod.getSpec().getSecurityContext().getRunAsGroup());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 3]
```diff
@@ -506,6 +519,60 @@ public void yamlOverrideVolume() throws Exception {
...
+        assertEquals(2000L, (Object) container.get().getSecurityContext().getRunAsUser());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(2000L, container.get().getSecurityContext().getRunAsUser());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 4]
```diff
@@ -506,6 +519,60 @@ public void yamlOverrideVolume() throws Exception {
...
+        assertEquals(2000L, (Object) container.get().getSecurityContext().getRunAsGroup());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(2000L, container.get().getSecurityContext().getRunAsGroup());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 5]
```diff
@@ -353,6 +364,8 @@ public void testInheritsFromWithYaml() throws Exception {
         assertEquals(new Quantity("1Gi"), jnlp.getResources().getLimits().get("memory"));
         assertEquals(new Quantity("100m"), jnlp.getResources().getRequests().get("cpu"));
         assertEquals(new Quantity("156Mi"), jnlp.getResources().getRequests().get("memory"));
+        assertEquals(1000L, (Object) jnlp.getSecurityContext().getRunAsUser());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(1000L, jnlp.getSecurityContext().getRunAsUser());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 6]
```diff
@@ -353,6 +364,8 @@ public void testInheritsFromWithYaml() throws Exception {
...
+        assertEquals(2000L, (Object) jnlp.getSecurityContext().getRunAsGroup());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(2000L, jnlp.getSecurityContext().getRunAsGroup());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 7]
```diff
@@ -266,6 +270,11 @@ private void validatePod(Pod pod, boolean fromYaml) {
             assertThat(jnlpMounts, containsInAnyOrder(volumeMounts));
         }
 
+        assertEquals(1000L, (Object) pod.getSpec().getSecurityContext().getRunAsUser());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(1000L, pod.getSpec().getSecurityContext().getRunAsUser());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 8]
```diff
@@ -266,6 +270,11 @@ private void validatePod(Pod pod, boolean fromYaml) {
...
+        assertEquals(1000L, (Object) pod.getSpec().getSecurityContext().getRunAsGroup());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(1000L, pod.getSpec().getSecurityContext().getRunAsGroup());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 9]
```diff
@@ -266,6 +270,11 @@ private void validatePod(Pod pod, boolean fromYaml) {
...
+        assertEquals(2000L, (Object) containers.get("busybox").getSecurityContext().getRunAsUser());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(2000L, containers.get("busybox").getSecurityContext().getRunAsUser());
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 10]
```diff
@@ -266,6 +270,11 @@ private void validatePod(Pod pod, boolean fromYaml) {
...
+        assertEquals(2000L, (Object) containers.get("busybox").getSecurityContext().getRunAsGroup());
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java]:
```suggestion
        assertEquals(2000L, containers.get("busybox").getSecurityContext().getRunAsGroup());
```

**Comment by georgelee1** [comment]
@Vlatombe Have adjusted the `assertEquals` object casting. Changed so that the expected is a `Long` object rather than a primitive. The compiler complains because of the overloading of `assertEquals` if using primitive `long` values and actual `Long` values.

**Comment by phillipuniverse** [comment]
This causes the following issue:

1. Upgrade from 120.1 -> 120.2. In 120.1, I obviously did not have anything for the user id or group id (no form elements)
2. On 120.2, make a change _only_ to the 'Raw YAML for the pod'. So, the 2 new form elements look like this:
    image
3. Save this change
4. When the form re-displays (like by refreshing the page after 'Apply' or  going back to the page after 'Save') then the form looks like this, where the values are non-null:
    image
5. This of course causes the following to be submitted when creating the Pod:
    image

> If the value is not set then the value will default to null which means that it is not set in the K8s deployment, which then means the user and group default to what ever the container image defined.

This part is not actually true. In fact, if the value is not set and then submitted, this defaults to 0. This is probably due to some default handling of numbers in the settings forms from Jenkins? This causes all my builds to run as the root user.

This means it is impossible to now create a pod that uses the defaults to the `USER` from the container.

In my case, the workaround was to ensure that all my k8s templates were set to the [same userId and groupId of the Jenkins Docker slave](https://github.com/jenkinsci/docker-slave/blob/master/Dockerfile#L29-L30), `1000:1000`.

**Comment by phillipuniverse** [comment]
1 more thing: I was utilizing a sidecar "docker-in-docker" container running with all of my build pods. This used to work just fine with just a `securityContext: privileged: "true"` like this:

image

However because I now have to set a `runAsUser` and `runAsGroup` at the global level, I have to override this for this specific container definition (`0:0` since I'm running as root). Final version looks like this:

image

**Comment by Vlatombe** [comment]
Thanks @phillipuniverse, I filed https://github.com/jenkinsci/kubernetes-plugin/pull/630 to fix this problem.

**Pull request merged**